libraries/Filter/HarmonicNotchFilter.cpp

```diff
@@ -0,0 +1,193 @@
+
+    // @Param: ATT
+    // @DisplayName: Harmonic Notch Filter attenuation
+    // @Description: Harmonic Notch Filter attenuation in dB
+    // @Range: 5 30
+    // @Units: dB
+    // @User: Advanced
+    AP_GROUPINFO("ATT", 4, HarmonicNotchFilterParams, _attenuation_dB, 15),
+
+    // @Param: HMNCS
+    // @DisplayName: Harmonic Notch Filter harmonics
+    // @Description: Bitmask of harmonic frequencies to apply Harmonic Notch Filter to. This option takes effect on the next reboot. A maximum of 3 harmonics can be used at any one time
+    // @Bitmask: 0:1st harmonic,1:2nd harmonic,2:3rd harmonic,3:4th hamronic,4:5th harmonic,5:6th harmonic,6:7th harmonic,7:8th harmonic
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("HMNCS", 5, HarmonicNotchFilterParams, _harmonics, 3),
+
+    // @Param: REF
+    // @DisplayName: Harmonic Notch Filter reference value
+    // @Description: Reference value associated with the specified frequency to facilitate frequency scaling of the Harmonic Notch Filter
+    // @User: Advanced
+    // @Range: 0.1 0.9
+    // @RebootRequired: True
+    AP_GROUPINFO("REF", 6, HarmonicNotchFilterParams, _reference, 0.1f),
+
+    AP_GROUPEND
+};
+
+/*
+  destroy all of the associated notch filters
+ */
+template <class T>
+HarmonicNotchFilter<T>::~HarmonicNotchFilter() {
+    delete[] _filters;
+    _num_filters = 0;
+}
+
+/*
+  initialise the associated filters with the provided shaping constraints
+  the constraints are used to determine attenuation (A) and quality (Q) factors for the filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, float bandwidth_hz, float attenuation_dB)
+{
+    if (_filters == nullptr) {
+        return;
+    }
+
+    _sample_freq_hz = sample_freq_hz;
+
+    // adjust the center frequency to be in the allowable range
+    center_freq_hz = constrain_float(center_freq_hz, bandwidth_hz * 0.52f, sample_freq_hz * 0.48f);
+
+    // calculate attenuation and quality from the shaping constraints
+    NotchFilter<T>::calculate_A_and_Q(center_freq_hz, bandwidth_hz, attenuation_dB, _A, _Q);
+
+    // initialize all the configured filters with the same A & Q and multiples of the center frequency
+    for (uint8_t i = 0, filt = 0; i < HNF_MAX_HARMONICS && filt < _num_filters; i++) {
+        if ((1U<<i) & _harmonics) {
+            _filters[filt++].init_with_A_and_Q(sample_freq_hz, center_freq_hz * (i+1), _A, _Q);
+        }
+    }
+    _initialised = true;
+}
+
+/*
+  allocate a collection of, at most HNF_MAX_FILTERS, notch filters to be managed by this harmonic notch filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::allocate_filters(uint8_t harmonics)
+{
+    for (uint8_t i = 0; i < HNF_MAX_HARMONICS && _num_filters < HNF_MAX_FILTERS; i++) {
+        if ((1U<<i) & harmonics) {
+            _num_filters++;
+        }
+    }
+    if (_num_filters > 0) {
+        _filters = new NotchFilter<T>[_num_filters];
```

Comment:
what happens if this allocation fails? We should at least show an error msg. Are all code paths that may use _filters ok with nullptr?

Reply:
Good catch. I will fix. Yes its safe because we never set _initialised if there is a nullptr.